File: simple-stack-example-mvvm-fragments/src/main/java/com/zhuinden/simplestackexamplemvvm/presentation/paths/taskdetail/TaskDetailKey.java
package com.zhuinden.simplestackexamplemvvm.presentation.paths.taskdetail;

import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;

import com.google.auto.value.AutoValue;
import com.zhuinden.simplestackexamplemvvm.R;
import com.zhuinden.simplestackexamplemvvm.application.BaseFragment;
import com.zhuinden.simplestackexamplemvvm.application.BaseKey;
import com.zhuinden.simplestackexamplemvvm.application.injection.Injection;
import com.zhuinden.simplestackexamplemvvm.core.viewmodels.ViewModelLifecycleHelper;

/**
 * Created by Zhuinden on 2017.07.26..
 */

@AutoValue
public abstract class TaskDetailKey
        extends BaseKey<TaskDetailViewModel> {
    public abstract String taskId();

    @Override
    protected boolean isFabVisible() {
        return true;
    }

    @Override
    protected void setupFab(Fragment fragment, FloatingActionButton fab) {
        fab.setImageResource(R.drawable.ic_edit);
        fab.setOnClickListener(v -> ((TaskDetailFragment) fragment).startEditTask());
    }

    @Override
    public int navigationViewId() {
        return 0;
    }

    @Override
    public boolean shouldShowUp() {
        return true;
    }

    @Override
    protected BaseFragment createFragment() {
        return new TaskDetailFragment();
    }

    @Override
    public ViewModelLifecycleHelper.ViewModelCreator<TaskDetailViewModel> getViewModelCreator() {
        return () -> Injection.get().taskDetailViewModel();
    }

    public static TaskDetailKey create(String taskId) {
        return new AutoValue_TaskDetailKey(taskId);
    }
}

File: samples/legacy-architecture-samples/simple-stack-example-mvvm-fragments/src/main/java/com/zhuinden/simplestackexamplemvvm/features/taskdetail/TaskDetailKey.kt
package com.zhuinden.simplestackexamplemvvm.features.taskdetail


import android.view.View
import androidx.fragment.app.Fragment
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.zhuinden.simplestack.ServiceBinder

import com.zhuinden.simplestackexamplemvvm.R
import com.zhuinden.simplestackexamplemvvm.application.BaseKey
import com.zhuinden.simplestackexamplemvvm.application.injection.Injector
import com.zhuinden.simplestackexamplemvvm.data.Task
import com.zhuinden.simplestackextensions.servicesktx.add
import kotlinx.android.parcel.Parcelize

/**
 * Created by Zhuinden on 2017.07.26..
 */
@Parcelize
data class TaskDetailKey(val task: Task) : BaseKey() {
    override fun bindServices(serviceBinder: ServiceBinder) {
        with(serviceBinder) {
            add(TaskDetailViewModel(
                Injector.get().tasksDataSource(),
                backstack,
                Injector.get().messageQueue(),
                getKey<TaskDetailKey>().task
            ))
        }
    }

    override val isFabVisible: Boolean
        get() = true

    override fun setupFab(fragment: Fragment, fab: FloatingActionButton) {
        fab.setImageResource(R.drawable.ic_edit)
        fab.setOnClickListener { v: View? -> (fragment as TaskDetailFragment).onEditTaskClicked() }
    }

    override fun navigationViewId(): Int {
        return 0
    }

    override fun shouldShowUp(): Boolean {
        return true
    }

    override fun instantiateFragment(): Fragment = TaskDetailFragment()
}